pypdf/_writer.py

```diff
@@ -1078,19 +1092,14 @@ def encrypt(
             V = 1
             rev = 2
             keylen = int(40 / 8)
-        secrets_generator = secrets.SystemRandom()
         P = permissions_flag
         O = ByteStringObject(_alg33(owner_password, user_password, rev, keylen))  # type: ignore[arg-type]  # noqa
-        ID_1 = ByteStringObject(md5((repr(time.time())).encode("utf8")).digest())
-        ID_2 = ByteStringObject(
-            md5((repr(secrets_generator.uniform(0, 1))).encode("utf8")).digest()
-        )
-        self._ID = ArrayObject((ID_1, ID_2))
+        self.generate_file_identifiers()
```

Comment:
shouldn't we just regenerate the ID_1/2 only when requested by the user ? like this, the ID will chang at each encryption. Is is expected?

Reply:
> like this, the ID will change at each encryption. Is is expected?

Yes, I was thinking the same. I actually think it's a bug.

To quote from the specs:

> External file references may be made more reliable by including a file identifier (PDF 1.1) in the file and using it
in addition to the normal platform-based file designation. Matching the identifier in the file reference with the
one in the file confirms whether the correct file was found.
File identifiers shall be defined by the optional ID entry in a PDF file’s trailer dictionary (see 7.5.5, “File Trailer”).
**The ID entry is optional but should be used**.

That is a weird combination. I think we should then just always write it? Maybe adding a flag that allows the user to disable writing the ID entry to make the produced PDF deterministic?

> The value of this entry shall be an array of two byte strings. The
**first byte string** shall be a permanent identifier based on the **contents of the file** at the time it was originally
created and **shall not change when the file is incrementally updated**.

To me that means it's pretty wrong what we do in multiple ways. However, it's not completely clear to me how to do it right.

A first step in the right direction would be not to overwrite the content of ID1 if it's present.

> The **second byte string** shall be a
changing identifier based on the **file’s contents at the time it was last updated**. When a file is first written, both
identifiers shall be set to the same value. If both identifiers match when a file reference is resolved, it is very
likely that the correct and unchanged file has been found. If only the first identifier matches, a different version
of the correct file has been found.

What I did in other projects is to calculate the MD5 hash of the complete file as an identifier for the file. Maybe we can "virtually" write the file .. .and maybe just using `_write_header` so that we skip the ID parameter in the trailer?
As a side-note: Should we rename `_write_header` to `_write_header_and_body`?

Reply:
I've updated the PR. Hopefully it still works / makes more sense :crossed_fingers: